### ISSUE TITLE
GoogleSignIn in Firebase for Flutter Codelab broken

### ISSUE THREAD

**Issue by smrtsmrf** [flutter/flutter#10688]
## Steps to Reproduce

I followed the steps in the Firebase for Flutter codelab (steps 5 & 6), but it doesn't work.

## Logs
```flutter run``` with an android emulator yields this result:
<img width="1628" alt="screen shot 2017-06-13 at 6 40 40 pm" src="https://user-images.githubusercontent.com/17756776/27115531-7cdb6198-5088-11e7-880c-11c30d151378.png">

```flutter run``` with iOS simulator yields this result:
<img width="1853" alt="screen shot 2017-06-13 at 10 37 12 pm" src="https://user-images.githubusercontent.com/17756776/27115596-fb325416-5088-11e7-8706-e23edfa892a1.png">
<img width="1838" alt="screen shot 2017-06-13 at 10 37 25 pm" src="https://user-images.githubusercontent.com/17756776/27115598-fdfc36f8-5088-11e7-8a05-31e23d533cb4.png">

## Flutter Doctor

<img width="1146" alt="screen shot 2017-06-13 at 10 38 30 pm" src="https://user-images.githubusercontent.com/17756776/27115617-1cec7208-5089-11e7-8665-465ee4781863.png">

## Other

Downgrading the plugin to v0.0.2 OR changing to the master channel of flutter via ```flutter channel master``` allowed the android run to actually get on the device, but the app crashes and won't actually start (despite the console saying it's ready to go and press "r" to hot reload - aka no indication of a problem).

**Comment by tvolkert** [comment]
The issue is that you're using a version of the plugin that was published today, yet it refers to a new class (`FlutterFragmentActivity`) in the Flutter engine that, while it was checked in 5 days ago (https://github.com/flutter/engine/pull/3757), does not yet exist on the `alpha` branch, which was tagged 12 days ago.

I'll release a new version to alpha tomorrow, which will partially resolve this.  Then we'll have to edit the codelab to instruct the developer to hand-edit their activity class (see https://github.com/flutter/flutter/issues/10690).

**Comment by sethladd** [comment]
Should we add a way for a plugin to declare a minimum version of the Flutter SDK? Then, the pub resolver might be able to match the plugin version to the version of Flutter SDK.

**Comment by tvolkert** [comment]
I think that exists (https://github.com/dart-lang/pub/issues/1432).  @nex3 what's the behavior in the following scenario:

1. plugin `P` version `0.0.1` declares a dependency on flutter SDK with no version constraint, like so:
```
dependencies:
  flutter:
    sdk: flutter
```

2. plugin `P` releases a new version `0.0.2` and declares a Flutter SDK version constraint:
```
dependencies:
  flutter:
    sdk: flutter

environment:
    flutter: ">=0.0.28"
```

3. Jane Developer has installed the Flutter SDK `0.0.27-dev`, and she runs a `pub upgrade` in her project that has an unconstrained dependency on `P`.

Is everything wired up correctly above to get the desired behavior, which is that `pub` satisfies the dependency on `P` with version `0.0.1`?

**Comment by tvolkert** [comment]
@collinjackson can you possibly update the codelab to reflect the need to manually update the activity to extend from `FlutterFragmentActivity`, or point me in the right direction (I hear special access is required)?

**Comment by nex3** [comment]
> Is everything wired up correctly above to get the desired behavior, which is that pub satisfies the dependency on P with version 0.0.1?

Yes. That said, I would generally caution you not to encourage users to publish packages with SDK constraints that have no version numbers, because by the same logic pub may fall back on selecting those package versions for users running Flutter 4.0.0 way down the line.

**Comment by collinjackson** [comment]
This was enough of a breaking change to Google Sign In that it probably deserved a more major version bump. @tvolkert what do you think?

I'll work with @LarkAscending to update the codelab with the `FlutterFragmentActivity` change. We should update the README as well. I'll submit a PR for that.

**Comment by tvolkert** [comment]
Sure, though at this point bumping the version again would just yield churn.

**Comment by collinjackson** [comment]
I just ran across https://developers.google.com/android/guides/api-client#manually_managed_connections and it actually doesn't seem so bad.

I think maybe we should just do that in the sign-in plugin, which would keep the codelab working.

We could still use `autoManage` if we detect that the app's Activity is a FragmentActivity.

**Comment by collinjackson** [comment]
It sounds like the agreed on plan is to switch the google_sign_in plugin to doing manual connection management if the provided activity isn't a FlutterFragmentActivity. This will fix #10690 as well. I am waiting for @tvolkert to finish doing the unification of internal and external google_sign_in plugins and then I will handle this.
